firmware/targets/f7/furi_hal/furi_hal_i2c.c

```diff
@@ -51,177 +51,248 @@ void furi_hal_i2c_release(FuriHalI2cBusHandle* handle) {
     furi_hal_power_insomnia_exit();
 }
 
-bool furi_hal_i2c_tx(
-    FuriHalI2cBusHandle* handle,
-    uint8_t address,
-    const uint8_t* data,
-    uint8_t size,
-    uint32_t timeout) {
-    furi_check(handle->bus->current_handle == handle);
-    furi_assert(timeout > 0);
+static bool wait_for_idle(I2C_TypeDef* i2c, I2CBegin begin, FuriHalCortexTimer timer) {
+    do {
+        if(furi_hal_cortex_timer_is_expired(timer)) {
+            return false;
+        }
+    } while(begin == I2CBeginStart && LL_I2C_IsActiveFlag_BUSY(i2c));
+    // Only check if the bus is busy if starting a new transaction, if not we already control the bus
+
+    return true;
+}
+
+static bool wait_for_end(I2C_TypeDef* i2c, I2CEnd end, FuriHalCortexTimer timer) {
+    // If ending the transaction with a stop condition, wait for it to be detected, otherwise wait for a transfer complete flag
+    bool wait_for_stop = end == I2CEndStop;
+    uint32_t end_mask = (wait_for_stop) ? I2C_ISR_STOPF : (I2C_ISR_TC | I2C_ISR_TCR);
+
+    while((i2c->ISR & end_mask) == 0) {
+        if(furi_hal_cortex_timer_is_expired(timer)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static uint32_t get_start_signal(I2CBegin begin, bool ten_bit_address, bool read) {
+    switch(begin) {
+    case I2CBeginRestart:
+        if(read) {
+            return ten_bit_address ? LL_I2C_GENERATE_RESTART_10BIT_READ :
+                                     LL_I2C_GENERATE_RESTART_7BIT_READ;
+        } else {
+            return ten_bit_address ? LL_I2C_GENERATE_RESTART_10BIT_WRITE :
+                                     LL_I2C_GENERATE_RESTART_7BIT_WRITE;
```

Comment:
This could be simplified, since both values are equal, depending on how close to ST's LL you want to stick.

Reply:
We'd rather hide anything related to ST LL inside.